File: backends/open-liberty-java/src/main/java/dev/stackverse/openliberty/MessageCatalog.java
package dev.stackverse.openliberty;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

@ApplicationScoped
public class MessageCatalog {
    @Inject RuntimeSupport runtime;

    String resolveLanguage(String lang, String acceptLanguage) {
        Set<String> supported = supportedLanguages();
        if (lang != null && supported.contains(lang)) return lang;
        for (LanguagePreference preference : parseAcceptLanguage(acceptLanguage)) {
            if (supported.contains(preference.code())) return preference.code();
        }
        return "en";
    }

    String firstParam(List<String> values) {
        return values == null || values.isEmpty() ? null : values.get(0);
    }

    String localize(String key, String language) {
        try (Connection connection = runtime.connection();
                PreparedStatement statement =
                        runtime.prepare(
                                connection,
                                """
             select text from messages
             where key = ? and language = any(?::text[])
             order by case when language = ? then 0 else 1 end
             limit 1
             """,
                                key,
                                new String[] {language, "en"},
                                language)) {
            try (ResultSet rs = statement.executeQuery()) {
                return rs.next() ? rs.getString("text") : key;
            }
        } catch (SQLException ex) {
            throw new RuntimeException(ex);
        }
    }

    Map<String, String> bundle(String language) throws SQLException {
        Map<String, String> messages = new LinkedHashMap<>();
        try (Connection connection = runtime.connection();
                PreparedStatement statement =
                        runtime.prepare(
                                connection,
                                """
             select key, language, text
             from messages
             where language = any(?::text[])
             order by key, case when language = ? then 0 else 1 end
             """,
                                new String[] {language, "en"},
                                language)) {
            try (ResultSet rs = statement.executeQuery()) {
                while (rs.next()) messages.putIfAbsent(rs.getString("key"), rs.getString("text"));
            }
        }
        return messages;
    }

    private Set<String> supportedLanguages() {
        try (Connection connection = runtime.connection();
                PreparedStatement statement =
                        connection.prepareStatement("select distinct language from messages")) {
            Set<String> result = new HashSet<>();
            try (ResultSet rs = statement.executeQuery()) {
                while (rs.next()) result.add(rs.getString("language"));
            }
            return result;
        } catch (SQLException ex) {
            throw new RuntimeException(ex);
        }
    }

    private static List<LanguagePreference> parseAcceptLanguage(String header) {
        if (header == null || header.isBlank()) return List.of();
        List<LanguagePreference> result = new ArrayList<>();
        int index = 0;
        for (String part : header.split(",")) {
            String[] pieces = part.trim().split(";");
            String code = pieces[0].trim().toLowerCase().split("-")[0];
            double quality = 1.0;
            for (int i = 1; i < pieces.length; i++) {
                String value = pieces[i].trim();
                if (value.startsWith("q=")) {
                    try {
                        quality = Double.parseDouble(value.substring(2));
                    } catch (NumberFormatException ignored) {
                        quality = 0;
                    }
                }
            }
            if (code.matches("^[a-z]{1,8}$"))
                result.add(new LanguagePreference(code, quality, index));
            index++;
        }
        result.sort(
                Comparator.comparingDouble(LanguagePreference::quality)
                        .reversed()
                        .thenComparingInt(LanguagePreference::index));
        return result;
    }
}

record LanguagePreference(String code, double quality, int index) {}

File: backends/ruby-rails/app/services/stackverse/message_catalog.rb
module Stackverse
  require "set"

  module MessageCatalog
    DEFAULT_LANGUAGE = "en"

    module_function

    def supported_languages
      Sql.query("select distinct language from messages").map { |row| row["language"] }.to_set
    end

    def resolve(lang, accept_language)
      supported = supported_languages
      return lang if lang.present? && supported.include?(lang)

      parse_accept_language(accept_language).each do |code|
        return code if supported.include?(code)
      end
      DEFAULT_LANGUAGE
    end

    def localize(key, language)
      languages = [ language, DEFAULT_LANGUAGE ].uniq
      row = Sql.one(<<~SQL.squish)
        select text from messages
        where key = #{Sql.quote(key)} and language = any(#{Sql.array(languages)})
        order by case when language = #{Sql.quote(language)} then 0 else 1 end
        limit 1
      SQL
      row ? row["text"] : key
    end

    def bundle(language)
      languages = [ language, DEFAULT_LANGUAGE ].uniq
      rows = Sql.query(<<~SQL.squish)
        select key, language, text from messages
        where language = any(#{Sql.array(languages)})
        order by key, case when language = #{Sql.quote(language)} then 0 else 1 end
      SQL
      rows.each_with_object({}) do |row, payload|
        payload[row["key"]] ||= row["text"]
      end
    end

    def parse_accept_language(header)
      return [] if header.blank?

      header.split(",").filter_map do |part|
        pieces = part.strip.split(";")
        tag = pieces.first.to_s.strip.downcase.split("-").first
        next unless tag.match?(/\A[a-z]{1,8}\z/)

        quality = pieces.drop(1).filter_map do |parameter|
          match = parameter.match(/\A\s*q=([0-9.]+)\s*\z/)
          match ? match[1].to_f : nil
        end.first || 1.0
        [ tag, quality.nan? ? 0.0 : quality ]
      end.sort_by { |_tag, quality| -quality }.map(&:first)
    end
  end
end
